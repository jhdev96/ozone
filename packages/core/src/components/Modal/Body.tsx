import React from 'react';
import styles from './index.module.scss';


interface Props {
  /**
   * Content that will rendered
   * inside of this component.
   */
  children?: React.ReactNode;
}

export default function Body({children}: Props): JSX.Element 
{
  return (
    <div className={styles.modalBody}>
      {children}
    </div>
  );
}

Body.displayName = 'ModalBody';