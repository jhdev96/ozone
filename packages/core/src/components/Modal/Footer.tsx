import React from "react";
import styles from './index.module.scss';


interface Props {
  /**
   * Content that will rendered
   * inside of this component.
   */
  children?: React.ReactNode;
}

export default function Footer({children}: Props): JSX.Element 
{
  return (
    <div className={styles.modalFooter}>
      {children}
    </div>
  );
}

Footer.displayName = 'ModalFooter';