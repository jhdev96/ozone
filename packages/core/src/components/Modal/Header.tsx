import React from 'react';
import styles from './index.module.scss';


interface HeaderProps {
  /**
   * Content that will rendered
   * inside of this component.
   */
  children?: React.ReactNode;
}

export default function Header({
  children
}: HeaderProps): JSX.Element 
{
  return (
    <div className={styles.modalHeader}>
      {children}
    </div>
  )
}

Header.displayName = 'ModalHeader';