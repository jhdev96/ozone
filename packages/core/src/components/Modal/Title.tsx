import React from 'react';
import styles from './index.module.scss';


interface TitleProps {
  /** 
   * The content for the title.
   */
  children?: React.ReactNode;
}

export default function Title({
  children
}: TitleProps): JSX.Element 
{
  return (
    <span className={styles.modalTitle}>
      {children}
    </span>
  )
}

Title.displayName = 'ModalTitle';