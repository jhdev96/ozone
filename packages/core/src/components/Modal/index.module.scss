@import '../../scss/root';


.modal {
  position: fixed;
  top: 0px;
  left: 0;
  z-index: -1;
  display: block !important;
  width: 100%;
  height: 100%;
  opacity: 0;
  overflow: hidden;
  outline: 0;
  overflow-x: hidden;
  overflow-y: auto;
  transition: 0.2s;
  background-color: #05151c68;

  .close {
    padding: 0;
    background-color: transparent;
    border: 0;
    float: right;
    line-height: 1;
    color: #000;
    text-shadow: 0 1px 0 #fff;
    opacity: .4;
    font-size: 1.2em;
    text-shadow: none !important;
    font-weight: 500;
    transition: 0.2s;
    cursor: pointer;
    box-shadow: none !important;
    outline: none !important;

    &:hover { opacity: 1; }
  }

  &Content {
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;
    pointer-events: auto;
    background-clip: padding-box;
    box-shadow: 1px 3px 25px 5px #00000017;
    border-radius: .3rem;
    outline: 0;
  }

  &Header {
    display: flex;
    flex-shrink: 0;
    align-items: center;
    justify-content: space-between;
    padding: 0.7rem 1rem;
    // border-bottom: 1px solid #dee2e6;
    border-top-left-radius: calc(.3rem - 1px);
    border-top-right-radius: calc(.3rem - 1px);
  }

  &Title {
    margin: 0.4em 0 !important;
    font-size: 1.2em;
    font-weight: 600;
  }

  &Footer {
    display: flex;
    flex-wrap: wrap;
    flex-shrink: 0;
    align-items: center;
    justify-content: flex-end;
    padding: .7rem;
    // border-top: 1px solid #dee2e6;
    border-bottom-right-radius: calc(.3rem - 1px);
    border-bottom-left-radius: calc(.3rem - 1px);
  }

  &Body {
    position: relative;
    flex: 1 1 auto;
    padding: 1rem;
  }

  &Dialog {
    position: relative;
    margin: .5rem;
    pointer-events: none;
    max-width: 500px;
    margin: 1.75rem auto;
    transform: translateY(50px);
    transition: transform .2s ease-out;
  }
}

/* Modifiers */
.visible {
  display: block;
  z-index: 1050;
  top: 0px !important;
  opacity: 1 !important;

  .modalDialog {
    transform: translateY(0px);
  }
}

.light {
  .modalContent {
    background-color: #fff;
  }
}

.dark {
  .modalContent {
    color: var(--gray-1);
    background-color: var(--black-4);
  }
  .modal__header {
    border-bottom: 1px solid var(--black-3);
  }
  .modal__footer {
    border-top: 1px solid var(--black-3);
  }
  .close { color: var(--gray-core); }
}

.fade { transition: opacity .15s linear; }