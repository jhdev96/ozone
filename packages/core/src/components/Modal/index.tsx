import React, { useState, useEffect, useRef } from 'react';
import { ThemeType } from '../../utils/theme';
import classNames from '../../utils/classNames';
import Header from './Header';
import Title from './Title';
import Body from './Body';
import Footer from './Footer';
import ClickManager from '../ClickManager';
import styles from './index.module.scss';


interface ModalProps {
  /** 
   * Give the modal a light or dark appearance. 
   */
  theme?: ThemeType;
  /** 
   * Indicates whether the modal should be visible. 
   */
  isVisible?: boolean;
  /** 
   * Callback to be invoked when the modal is closed. 
   */
  onClose?: () => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The content that will be rendered inside the modal. 
   */
  children?: React.ReactNode;
}

export default function Modal({
  theme     = 'light',
  isVisible = false,
  onClose   = () => {},
  dataTestId,
  children
}: ModalProps): JSX.Element 
{
  const [visible, setVisible] = useState(false);
  const wrapperRef = useRef<HTMLDivElement>();

  useEffect(() => {
    setVisible(isVisible);
  }, [isVisible]);

  const handleClose = () => {
    if (visible) {
      onClose();
      setVisible(false);
    }
  }

  return (
    <div 
      className={classNames(
        styles.modal, 
        visible && styles.visible,
        styles[theme]
      )}
      data-testid={dataTestId}
    >
      <ClickManager 
        onClick={handleClose} 
        ref={wrapperRef}
      >
        <div className={styles.modalDialog}>
          <div className={styles.modalContent}>
            {children}
          </div>
        </div>
      </ClickManager>
    </div>
  );
}

Modal.Header = Header;
Modal.Title = Title;
Modal.Body = Body;
Modal.Footer = Footer;

Modal.displayName = 'Modal';