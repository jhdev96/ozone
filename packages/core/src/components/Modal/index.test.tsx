import React from "react";
import { render, screen, fireEvent } from '@testing-library/react';
import Modal from '.';


describe('<Modal />', () => {
  it('renders a hidden modal by default', () => {
    render(
      <Modal dataTestId="test">
        <Modal.Body>Text</Modal.Body>
      </Modal>
    );

    expect(screen.getByTestId("test"))
      .not.toHaveClass('visible');
  });

  it('renders a modal with a header', () => {
    render(
      <Modal dataTestId="test">
        <Modal.Header>Title</Modal.Header>
      </Modal>
    );

    expect(screen.getByTestId("test"))
      .toMatchSnapshot();
  });

  it('renders a modal with a title', () => {
    render(
      <Modal dataTestId="test">
        <Modal.Header>
          <Modal.Title>Title</Modal.Title>
        </Modal.Header>
      </Modal>
    );

    expect(screen.getByTestId("test"))
      .toMatchSnapshot();
  });

  it('renders a visible modal when `isVisible` is set', () => {
    render(
      <Modal dataTestId="test" isVisible>
        <Modal.Body>Text</Modal.Body>
      </Modal>
    );

    expect(screen.getByTestId("test"))
      .toHaveClass('visible');
  });

  it('hides the modal when the user clicks outside the dialog', () => {
    render(
      <Modal dataTestId="test" isVisible>
        <Modal.Body>Text</Modal.Body>
      </Modal>
    );

    const modalOverlay = document.querySelector('.modal');
    fireEvent.click(modalOverlay!);

    expect(screen.getByTestId("test"))
      .not.toHaveClass('visible');
  });

  it('leaves the modal open when the user clicks inside the modal', () => {
    render(
      <Modal dataTestId="test" isVisible>
        <Modal.Body>Text</Modal.Body>
      </Modal>
    );

    const modalContent = document.querySelector('.modalBody');
    fireEvent.click(modalContent!);

    expect(screen.getByTestId("test"))
      .toHaveClass('visible');
  });

  it('calls `onClose` when the modal is closed', () => {
    const onClose = jest.fn();

    render(
      <Modal 
        onClose={onClose} 
        dataTestId="test" 
        isVisible
      >
        <Modal.Body>Text</Modal.Body>
      </Modal>
    );

    const modalOverlay = document.querySelector('.modal');
    fireEvent.click(modalOverlay!);

    expect(onClose).toBeCalled;
  });
});