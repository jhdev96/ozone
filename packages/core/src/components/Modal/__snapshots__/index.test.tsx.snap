// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders a modal with a header 1`] = `
<div
  class="modal light"
  data-testid="test"
>
  <div>
    <div
      class="modalDialog"
    >
      <div
        class="modalContent"
      >
        <div
          class="modalHeader"
        >
          Title
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Modal /> renders a modal with a title 1`] = `
<div
  class="modal light"
  data-testid="test"
>
  <div>
    <div
      class="modalDialog"
    >
      <div
        class="modalContent"
      >
        <div
          class="modalHeader"
        >
          <span
            class="modalTitle"
          >
            Title
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
